adafruit_debouncer.py

```diff
@@ -0,0 +1,131 @@
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+"""
+`adafruit_debouncer`
+====================================================
+
+Debounces an arbitrary predicate function (typically created as a lambda) of 0 arguments.
+Since a very common use is debouncing a digital input pin, the initializer accepts a pin number
+ instead of a lambda.
+
+* Author(s): Dave Astels
+
+Implementation Notes
+--------------------
+
+**Hardware:**
+
+**Software and Dependencies:**
+
+* Adafruit CircuitPython firmware for the supported boards:
+  https://github.com/adafruit/circuitpython/releases
+"""
+
+# imports
+
+__version__ = "0.0.0-auto.0"
+__repo__ = "https://github.com/adafruit/Adafruit_CircuitPython_Debouncer.git"
+
+import time
+import digitalio
+import microcontroller
+
+class Debouncer(object):
+    """Debounce an input pin or an arbitrary predicate"""
+
+    DEBOUNCED_STATE = 0x01
+    UNSTABLE_STATE = 0x02
+    CHANGED_STATE = 0x04
+
+
+    def __init__(self, pin_or_predicate, mode=None, interval=0.010):
+        """Make am instance.
+           :param int/function pin_or_predicate: the pin (from board) to debounce
+           :param int mode: digitalio.Pull.UP or .DOWN (default is no pull up/down)
+           :param int interval: bounce threshold in seconds (default is 0.010, i.e. 10 milliseconds)
+        """
+        self.state = 0x00
+        if isinstance(pin_or_predicate, microcontroller.Pin):
+            pin = digitalio.DigitalInOut(pin_or_predicate)
+            pin.direction = digitalio.Direction.INPUT
+            if mode is not None:
+                pin.pull = mode
+            self.function = lambda: pin.value
+        else:
+            self.function = pin_or_predicate
+        if self.function():
+            self.__set_state(Debouncer.DEBOUNCED_STATE | Debouncer.UNSTABLE_STATE)
+        self.previous_time = 0
+        if interval is None:
+            self.interval = 0.010
+        else:
+            self.interval = interval
+
+
+    def __set_state(self, bits):
```

Comment:
Not so much a request for a change; more of an "enquiring mind wants to know".

Why the choice of double underscores to prefix these methods?

Reply:
I didn't know better when I wrote the code originally, months ago.  Changed to single underscore, which I believe is the correct way to denote a member as "private".